### PR TITLE
JS task

### PR DESCRIPTION
**Pull request by mike3293**

### CODE REVIEW

File: js-task/FP_functions.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
+const fp = {};
+(function(fp) {
```
Comment (ilyahumbar) [inline js-task/FP_functions.js]: Why not ES6 modules?

File: js-task/FP_functions.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    // bind()
+    fp.partial = function(...args) {
+        const foo = args.pop();
```
Comment (ilyahumbar) [inline js-task/FP_functions.js]: Be careful with naming. Currently this name doesn't say anything about what it stores. Rename (and all other `foo` references as well)

File: js-task/FP_functions.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            } else {
+                let result = foo(n);
+                storage[n] = result;
+
+                return result;
+            }
+        }
+    }
+}(fp));
+module.exports = fp;
```
Comment (ilyahumbar) [inline js-task/FP_functions.js]: You will get an error when trying to use your library in browser

File: js-task/FP_functions.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            if (args.length >= arity) {
+                return foo(...args);
+            } else {
+                return (a) => {
+                    args.push(a);
+                    return innerFunction(...args);
+                }
+            }
+        }
+    }
```
Comment (ilyahumbar) [inline js-task/FP_functions.js:34]: Test case which is not working properly:

```
const a = function (a, b, c, d) {
  return a + b + c + d;
}
const b = fp.curry(a);
const c = b(1);
const d = c(2);

c(4); /*should return function. That's correct*/
d(5); /*should also return function. But returns 10 in your code*/
```

File: js-task/FP_functions.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    //reduce()
+    fp.fold = function(array, callback, initialValue) {
+        if (typeof callback !== 'function') {
+            throw new TypeError(callback + ' is not a function');
+        }
+        if (!Array.isArray(array)) {
+            throw new TypeError(array + ' is not an Array object');
+        }
+        if (array.length == 0) {
+            throw new TypeError('array is empty');
```
Comment (ilyahumbar) [inline js-task/FP_functions.js]: The behavior should be the same as for default `reduce` function. It's not throwing exception when initial value is provided

File: js-task/FP_functions.js [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            accum = array[0];
+            i = 1;
+        } else if (!isNaN(initialValue)) {
+            accum = initialValue;
+            i = 0;
+        } else {
+            throw new TypeError(initialValue + ' is not a number');
+        }
+
+        for (; i < array.length; i++) {
```
Comment (ilyahumbar) [inline js-task/FP_functions.js]: Implement without loops. Use functional techniques

Comment (ilyahumbar) [reply js-task/FP_functions.js]: Same for `fp.unfold`, `fp.map`, `fp.filter`, `fp.first`,

File: js-task/FP_functions.js [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            throw new TypeError(foo + ' is not a function');
+        }
+        const storage = {};
+
+        return function(n) {
+            if (n in storage) {
+                return storage[n];
+            } else {
+                let result = foo(n);
+                storage[n] = result;
```
Comment (ilyahumbar) [inline js-task/FP_functions.js]: What would be if `n` is object? What if 'n' is undefined?

File: js-task/test/test_01.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+QUnit.test("partial() normal work", function(assert) {
+    assert.equal(fp.partial(1, sum)(2, 3), 6, "partial(1, sum)(2, 3) = 6");
+});
+QUnit.test("partial() work without part of parameters", function(assert) {
+    assert.equal(fp.partial(2, 3, sum)(), 5, "partial(2, 3, sum)() = 5");
+});
+QUnit.test("partial() get only function", function(assert) {
+    assert.equal(fp.partial(sum)(), 0, "partial(sum)() = 0 (sum)");
+});
+QUnit.test("partial() throw", function(assert) {
```
Comment (ilyahumbar) [inline js-task/test/test_01.js]: Name is not very descriptive. Throw what and in what conditions?
Same for other tests

File: js-task/test/test_10.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+const fp = require("../FP_functions.js");
+
+QUnit.test("lazy() work", function(assert) {
+    const sum = (a, b) => { return a + b; }
+
+    const lazyFold = fp.lazy(fp.fold, [1, 2, 3], sum, 4);
+
+    assert.ok(typeof lazyFold === "function", "lazyFold is a function");
+    assert.equal(lazyFold(), fp.fold([1, 2, 3], sum, 4), "result of lazyFold() and fold([1, 2, 3], sum, 4) is the same");
```
Comment (ilyahumbar) [inline js-task/test/test_10.js]: It also should test that given function is called only once.
e.g. 
```
lazyFold();
lazyFold();
lazyFold();
// add test here that initial function was called only one time
```

File: js-task/FP_functions.js (hunk 1 above)
Comment (mike3293) [reply js-task/FP_functions.js]: Changed to ES6 modules

File: js-task/FP_functions.js (hunk 7 above)
Comment (mike3293) [reply js-task/FP_functions.js]: Added check for invalid args

File: js-task/FP_functions.js (hunk 4 above)
Comment (mike3293) [reply js-task/FP_functions.js:34]: Edited the test

Comment (ilyahumbar) [reply js-task/FP_functions.js:34]: Still an issue in your code. Nothing's fixed. More over, if you call `c(4)` and then `c(5)` result will be the same.

**Pull request merged**